anndata/_core/anndata.py

```diff
@@ -592,16 +592,25 @@ def _init_as_actual(
         # layers
         self._layers = Layers(self, layers)
 
-    def __sizeof__(self, show_stratified=None) -> int:
+    def __sizeof__(self, show_stratified=None, with_disk: bool = False) -> int:
         def get_size(X):
-            if issparse(X):
-                X_csr = csr_matrix(X)
-                return X_csr.data.nbytes + X_csr.indptr.nbytes + X_csr.indices.nbytes
+            def cs_to_bytes(X):
+                return X.data.nbytes + X.indptr.nbytes + X.indices.nbytes
+
+            if not with_disk:
+                return X.__sizeof__()
```

Comment:
What if `with_disk=False` and `isinstance(X, (sparse.csr_matrix, sparse.csc_matrix))`?

Reply:
Hello,
Would you mind elaborating a little bit ?
In particular, what other objects should I exclude in the case of `with_disk=False` ?

Reply:
> What if `with_disk=False` and `isinstance(X, (sparse.csr_matrix, sparse.csc_matrix))`?

In this case, I thought that in case of `with_disk=False` then we return the `sparse.csr_matrix`/`sparse.csc_matrix` `__sizeof__()` whose implementation is `scipy.sparse` responsibility.

Reply:
>  In particular, what other objects should I exclude in the case of with_disk=False ?

See above:

> `with_disk=False` → all in-memory structures, sparse or dense.

So csr_matrix, csc_matrix, dataframe (if we support that here yet), and numpy array

> we return the sparse.csr_matrix/sparse.csc_matrix __sizeof__() whose implementation is scipy.sparse responsibility.

That returns `24`, that’s why we have a code branch for it.

Reply:
If I ask for some clarification it's because it wasn't obvious from my first reading of your specifications. I'm currently clueless on how to differentiate between `everything` and `all in-memory structures, sparse or dense`.


> That returns 24, that’s why we have a code branch for it.
Then isn't it what we're supposed to return as well? 


Best,

Reply:
Hello,
Thanks for clarifying, I've adjusted the function so that `with_disk` only matters when X is an `h5py.Dataset`, which is the only difference I've noted compared to my previous implementation.

Now on checking the second use case, when manipulating a backed AnnData I've encountered another bug:
(Note: for testing purposes I've turned off the size conversion from B to MB when using `show_stratified=True`)

```python 
>>> from anndata import AnnData
>>> import numpy as np
>>> from scipy.sparse import issparse, csr_matrix, csc_matrix
>>> DS = AnnData(np.zeros((100,100)))
>>> DS.__sizeof__()
100244
>>> DS.__sizeof__(show_stratified=True)
Size of X      : 24.00
Size of .obs   : 10058.00
Size of .var   : 10058.00
20140
>>> DS.filename = 'backed.h5ad'
>>> DS_b = AnnData(filename='backed.h5ad', filemode='r')
>>> DS_b.__sizeof__()
240
>>> DS_b.__sizeof__(with_disk=True, show_stratified=True)
Size of X      : 80000.00
Size of .obs   : 108.00
Size of .var   : 108.00
80216
>>> DS.obs
Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]

[100 rows x 0 columns]
>>> DS_b.obs
Empty DataFrame
Columns: []
Index: []
>>> DS_mem = DS_b.to_memory()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ejodry/Repositories/anndata/anndata/_core/anndata.py", line 1579, in to_memory
    return AnnData(**new)
           ^^^^^^^^^^^^^^
  File "/home/ejodry/Repositories/anndata/anndata/_core/anndata.py", line 362, in __init__
    self._init_as_actual(
  File "/home/ejodry/Repositories/anndata/anndata/_core/anndata.py", line 545, in _init_as_actual
    self._obs = _gen_dataframe(
                ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ejodry/Repositories/anndata/anndata/_core/anndata.py", line 180, in _gen_dataframe_df
    raise _mk_df_error(source, attr, length, len(anno))
ValueError: Observations annot. `obs` must have as many rows as `X` has rows (100), but has 0 rows.
```